Add pickWithDefault and deprecate pickAll

olsonpm opened the pull request:
- Fixes #1876

davidchambers commented on src/pickWithDefault.js:
```diff
@@ -0,0 +1,36 @@
+ *
+ * @func
+ * @memberOf R
+ * @since v0.22.2
```
Leave out `@since`. The build script will take care of it. :)

olsonpm replied:
Ah sorry - I read the contributing guidelines after having added it and forgot.

davidchambers commented on src/pickWithDefault.js:
```diff
@@ -0,0 +1,36 @@
+ * @memberOf R
+ * @since v0.22.2
+ * @category Object
+ * @sig [k] -> * -> {k: v} -> {k: v}
```
I'd like to see:

```haskell
pickWithDefault :: Array String -> a -> StrMap a -> StrMap a
```

We need to start using [`StrMap`][1] in signatures where appropriate if we're to express constraints in type signatures such as this one.


[1]: https://github.com/sanctuary-js/sanctuary-def#strmap

olsonpm replied:
Gotcha - I tried keeping it exactly like pickAll

olsonpm replied:
@davidchambers - Is there an example signature I can follow?  I don't see 'Array String' anywhere but i do see ['[String]'](http://ramdajs.com/0.22.1/docs/#mapObjIndexed).  I was about to copy paste exactly what you put until I checked the website.

davidchambers replied:
> I tried keeping it exactly like pickAll

Yep. Our signatures need some attention. There's an argument for keeping things consistent until we update *all* the signatures, but that's a daunting prospect. I'm happy to make small improvements here and there.

> I don't see 'Array String' anywhere but i do see ['[String]'](http://ramdajs.com/0.22.1/docs/#mapObjIndexed).

I switched to using `Array a` rather than `[a]` in Sanctuary, but Ramda uses the latter. For Ramda I'd like:

```haskell
pickWithDefault :: [String] -> a -> StrMap a -> StrMap a
```

As far as I'm aware we're not yet using `StrMap` in any of our signatures. I'm interested to know whether @buzzdecafe and @CrossEye are interested in adopting it.

olsonpm replied:
Thanks much for the clarification.  I'll make that change and push the items you've noted.  They can give feedback then.

Yeah my only concern with Sanctuary constructs in ramda signatures is that the front-page documentation should be updated to make users aware of the strong ties.

davidchambers replied:
`StrMap` is not really a Sanctuary construct. In fact, I got the idea from [PureScript][1].

JavaScript objects are used in two distinct ways: as records (with known fields of known—possibility heterogeneous—types), and as dictionaries (with string keys and homogeneous values). In the latter case we need a unary type so we can indicate the type of the dictionary's values. This is what `StrMap` provides.

I agree that if we're to mention `StrMap` in type signatures we should document it in Ramda itself.


[1]: https://pursuit.purescript.org/packages/purescript-maps/1.1.0/docs/Data.StrMap

davidchambers commented on src/pickWithDefault.js:
```diff
@@ -0,0 +1,36 @@
+ * @category Object
+ * @sig [k] -> * -> {k: v} -> {k: v}
+ * @param {Array} names an array of String property names to copy onto a new object
+ * @param * default value when encountering undefined
```
This is misleading. The default value is only used for missing keys.

davidchambers commented on src/pickWithDefault.js:
```diff
@@ -0,0 +1,36 @@
+  while (idx < names.length) {
+    result[names[idx]] = (names[idx] in obj)
+      ?  obj[names[idx]]
+      : def
```
I'd prefer this formatting:

```javascript
result[names[idx]] = names[idx] in obj ? obj[names[idx]] : def;
```

olsonpm replied:
Any chance at getting an autoformatter?

davidchambers replied:
I believe ESLint can automatically fix some problems, but last time I checked only a fraction of the rules we would like to enforce are "fixable".

olsonpm replied:
Yeah I'd have to see the status of js-beautify - it does quite a bit.

davidchambers commented on test/pickWithDefault.js:
```diff
@@ -0,0 +1,20 @@
+  it('is curried', function() {
+    var copyAdefaultG = R.pickWithDefault(['a', 'g']);
+    eq(copyAdefaultG(null, obj), {a: 1, g: null});
+  });
```
We could remove this test. All functions in the library are curried, and we already have a test which asserts this.

olsonpm replied:
Yeah I wondered about that.  Again I just copied from pickAll but will remove

davidchambers commented on src/pickAll.js:
```diff
@@ -14,6 +14,7 @@ var _curry2 = require('./internal/_curry2');
  * @param {Object} obj The object to copy from
  * @return {Object} A new object with only properties from `names` on it.
  * @see R.pick
+ * @deprecated since v0.23.0
```
Let's update __src/project.js__ to use `pickWithDefault` rather than `pickAll`.

olsonpm commented:
Alright - I'm pretty sure I addressed all your comments.  I guess we just await feedback on the signature change.  I'm assuming they'll be on board with making incremental modifications to the signatures, just not sure about including sanctuary constructs in the signature.

davidchambers commented on src/project.js:
```diff
@@ -23,4 +24,4 @@ var useWith = require('./useWith');
  *      R.project(['name', 'grade'], kids); //=> [{name: 'Abby', grade: 2}, {name: 'Fred', grade: 7}]
  */
-module.exports = useWith(_map, [pickAll, identity]); // passing `identity` gives correct arity
+module.exports = useWith(_map, [pickWithDefault(__, undefined), identity]); // passing `identity` gives correct arity
```
This suggests that the default value should be the first argument rather than the second.

olsonpm replied:
I mean, there's no real objective way to know until you start consuming the api in practice.

olsonpm replied:
Pushed the change - you are probably right.  I'm just assuming the default value is less likely to vary based on input.

CrossEye commented:
LGTM

:herb:

ChrisPenner commented on src/pickWithDefault.js:
```diff
@@ -0,0 +1,32 @@
+ * @example
+ *
+ *      R.pickWithDefault(['a', 'd'], {a: 1, b: 2, c: 3, d: 4}); //=> {a: 1, d: 4}
+ *      R.pickWithDefault(['a', 'e', 'f'], null, {a: 1, b: 2, c: 3, d: 4}); //=> {a: 1, e: null, f: null}
```
These examples are incorrect, In the former there's no default argument, in the latter the default argument should come *before* the paths.

oscar-b commented:
What's the status on `pickWithDefault`?

CrossEye commented:
@oscar-b:

> What's the status on `pickWithDefault`?

I think it simply got dropped.  I think it was a reasonable idea.  If you feel like reviving it by fixing the conflicts, I would certainly consider merging it.

CrossEye commented:
Too long abandoned.  Can revisit if anyone wants to revive it.
